Sources/XMTPiOS/Contacts.swift

```diff
@@ -70,75 +75,78 @@ public class ConsentList {
+				_ = allowGroup(groupId: groupId)
 			}
 
 			for groupId in preference.denyGroup.groupIds {
-				_ = consentList.denyGroup(groupId: groupId)
+				_ = denyGroup(groupId: groupId)
 			}
 		}
 
-		return consentList
+        return Array(entries.values)
 	}
 
-	func publish(entry: ConsentListEntry) async throws {
-		guard let identifier = identifier else {
-			throw ContactError.invalidIdentifier
-		}
-
-		var payload = PrivatePreferencesAction()
-		switch entry.entryType {
-
-		case .address:
-			switch entry.consentType {
-			case .allowed:
-				payload.allowAddress.walletAddresses = [entry.value]
-			case .denied:
-				payload.denyAddress.walletAddresses = [entry.value]
-			case .unknown:
-				payload.messageType = nil
-			}
-
-		case .groupId:
-			switch entry.consentType {
-			case .allowed:
-				if let valueData = entry.value.data(using: .utf8) {
-					payload.allowGroup.groupIds = [valueData]
-				}
-			case .denied:
-				if let valueData = entry.value.data(using: .utf8) {
-					payload.denyGroup.groupIds = [valueData]
-				}
-			case .unknown:
-				payload.messageType = nil
-			}
-		}
-
-		let message = try LibXMTP.userPreferencesEncrypt(
-			publicKey: publicKey,
-			privateKey: privateKey,
-			message: payload.serializedData()
-		)
-
-		let envelope = Envelope(
-			topic: Topic.preferenceList(identifier),
-			timestamp: Date(),
-			message: Data(message)
-		)
-
-		try await client.publish(envelopes: [envelope])
-	}
+    func publish(entries: [ConsentListEntry]) async throws {
+      guard let identifier = identifier else {
+        throw ContactError.invalidIdentifier
+      }
+      var payload = PrivatePreferencesAction()
+
+      for entry in entries {
+        switch entry.entryType {
+
+    	    case .address:
+    	      switch entry.consentType {
+    	      case .allowed:
+    	          payload.allowAddress.walletAddresses.append(entry.value)
+    	      case .denied:
+    	          payload.denyAddress.walletAddresses.append(entry.value)
+    	      case .unknown:
+    	          payload.messageType = nil
+    	      }
+
+    			case .groupId:
+    	    	switch entry.consentType {
+    	    	case .allowed:
+    	    	    if let valueData = entry.value.data(using: .utf8) {
+    	    	        payload.allowGroup.groupIds.append(valueData)
+    	    	    }
+    	    	case .denied:
+    	    	    if let valueData = entry.value.data(using: .utf8) {
+    	    	            payload.denyGroup.groupIds.append(valueData)
+    	    	    }
+    	    	case .unknown:
+    	    	    payload.messageType = nil
+    	    }
+    	}
+
+    }
+
+
+    let message = try LibXMTP.userPreferencesEncrypt(
+        publicKey: publicKey,
+        privateKey: privateKey,
+        message: payload.serializedData()
+    )
+
+    let envelope = Envelope(
+        topic: Topic.preferenceList(identifier),
+        timestamp: Date(),
+        message: Data(message)
+    )
+
+    try await client.publish(envelopes: [envelope])
+  }
```

Comment:
I think this logic is a little incorrect. I think we want to pass in an array of envelopes. Like this where the envelopes are returned from the loop.

```suggestion
    func publish(entries: [ConsentListEntry]) async throws {
      guard let identifier = identifier else {
        throw ContactError.invalidIdentifier
      }
      var payload = PrivatePreferencesAction()

      let envelopes = entries.map { entry ->
        switch entry.entryType {

    	    case .address:
    	      switch entry.consentType {
    	      case .allowed:
    	          payload.allowAddress.walletAddresses.append(entry.value)
    	      case .denied:
    	          payload.denyAddress.walletAddresses.append(entry.value)
    	      case .unknown:
    	          payload.messageType = nil
    	      }

    			case .groupId:
    	    	switch entry.consentType {
    	    	case .allowed:
    	    	    if let valueData = entry.value.data(using: .utf8) {
    	    	        payload.allowGroup.groupIds.append(valueData)
    	    	    }
    	    	case .denied:
    	    	    if let valueData = entry.value.data(using: .utf8) {
    	    	            payload.denyGroup.groupIds.append(valueData)
    	    	    }
    	    	case .unknown:
    	    	    payload.messageType = nil
    	    }
    	}
    	
        let message = try LibXMTP.userPreferencesEncrypt(
          publicKey: publicKey,
          privateKey: privateKey,
          message: payload.serializedData()
        )

        Envelope(
          topic: Topic.preferenceList(identifier),
          timestamp: Date(),
          message: Data(message)
         )
    }

    try await client.publish(envelopes: envelopes)
  }
```

Reply:
@nplasterer I can update to multiple envelopes, but this payload supports multiple addresses being added (not sure which is actually ideal from network/device performance)

Reply:
I'm not sure I understand the question?

Reply:
Ooo you're using append instead of add. 🤔

Reply:
> Ooo you're using append instead of add. 🤔

Swift :)

> I'm not sure if we tested this case. I think if you just add a test for calling `deny(addresses: [String]) ` with a list of addresses. And make sure that all the addresses are denied. This should be fine with just one envelope. 🤔

Added a new test case to verify those